### PR TITLE
Layers nested query rework

### PR DESCRIPTION
**Pull request by aldebaran30701**
### Description of the Change
NOTE: This PR contains state changes from the Layers View State branch also.

Rework query language used by layers view. 
New language allows for high levels of nesting, escaping of characters.
Added test cases for the `QueryEvaluator.java` and `UpdateElementBitmaskPlugin.java`
Draw mode now adds an element to the default layer 1 as well as the layers that are currently selected.

### Alternate Designs
JEXL - Implementation did not meet criteria as certain characters are reserved and not escapable. Eg. '#'
Split by ':' was very messy when writing queries.

### Why Should This Be In Core?
It adds functionality to the layers view and also fixes some layers view bugs

### Benefits
Layers view state is kept and storage of layers on graph will allow users to spend less time re configuring views.

### Possible Drawbacks
Needs more storage for saving a graph - minimal.

### Verification Process
Unit tests - for the classes with them implemented.
Various UI calls to test. 
Layers view state testing outlined in PR #454 
Bugs fixed with relation to invisibility were attempted to be replicated with many new graphs and many additions / switches of layers.

### Applicable Issues
Layers view issue #223  - Layers view state #454

### CODE REVIEW

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/state/AnalyticViewState.java [hunk 1]
```diff
@@ -90,17 +98,55 @@ public void clearAnalyticQuestions() {
...
+        public Set<Class<? extends SchemaConcept>> getParents() {
+            final Set<Class<? extends SchemaConcept>> parentSet = new HashSet<>();
+            parentSet.add(ConstellationViewsConcept.class);
+            return Collections.unmodifiableSet(parentSet);
+        }
+
+        public static class MetaAttribute {
+
+            private MetaAttribute() {
+                throw new IllegalStateException("Utility class");
```
Comment (antares1470) [inline CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/state/AnalyticViewState.java:137]: The concept classes aren't really utility classes (at least by our definitions anyway). We usually mark the sonar rule that comes up as a false positive in these cases. I won't say it's a bad thing to include the private constructor but maybe change the string in the exception (something like "Concept class)").

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java [hunk 2]
```diff
@@ -109,47 +118,95 @@ public LayersViewPane(final LayersViewController controller) {
...
-        visibilityCheckBox.setPadding(new Insets(0, 30, 0, 15));
-        visibilityCheckBox.selectedProperty().addListener((observable, oldValue, newValue) -> {
+        final Node visibilityCheckBox = new CheckBox();
+        
+        ((CheckBox) visibilityCheckBox).setMinWidth(60);
+        ((CheckBox) visibilityCheckBox).setPadding(new Insets(0, 30, 0, 15));
+        ((CheckBox) visibilityCheckBox).setSelected(checkBoxSelected);
+        visibilityCheckBox.setOnMouseClicked(e -> {
+            Node source = (Node) e.getSource();
+            LayerDescription layer = layers.get(GridPane.getRowIndex(source)-1);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java:142]: source and layer can both be made final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java [hunk 3]
```diff
@@ -109,47 +118,95 @@ public LayersViewPane(final LayersViewController controller) {
...
             controller.execute();
+            controller.updateState(true);
+        });
+        
+        final Node queryTextArea = new TextArea();
+        ((TextArea) queryTextArea).setPrefRowCount(1);
+        ((TextArea) queryTextArea).setText(query);
+        ((TextArea) queryTextArea).focusedProperty().addListener((observable, oldVal, newVal) -> {
+            if(!newVal){
+                LayerDescription layer = layers.get(currentIndex-1);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java:154]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java [hunk 4]
```diff
@@ -109,47 +118,95 @@ public LayersViewPane(final LayersViewController controller) {
...
+                controller.updateState(true);
+            }
+        });
+        
+        final Node descriptionTextArea = new TextArea();
+        ((TextArea) descriptionTextArea).setPrefRowCount(1);
+        ((TextArea) descriptionTextArea).setText(description);
+        ((TextArea) descriptionTextArea).focusedProperty().addListener((observable, oldVal, newVal) -> {
+            if(!newVal){
+                LayerDescription layer = layers.get(currentIndex-1);
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java:165]: final

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/state/LayersViewState.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        public Set<Class<? extends SchemaConcept>> getParents() {
+            final Set<Class<? extends SchemaConcept>> parentSet = new HashSet<>();
+            parentSet.add(SchemaConcept.ConstellationViewsConcept.class);
+            return Collections.unmodifiableSet(parentSet);
+        }
+
+        public static class MetaAttribute {
+
+            private MetaAttribute() {
+                throw new IllegalStateException("Utility class");
```
Comment (antares1470) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/state/LayersViewState.java:98]: Similar comment as above regarding utility class

File: CoreUtilities/src/au/gov/asd/tac/constellation/utilities/query/QueryEvaluator.java [hunk 6]
```diff
@@ -17,108 +17,221 @@
...
+                boolean moreToEscape = true;
+                String trimmedToken = token;
+
+                while (trimmedToken.contains("\\") && moreToEscape) {
+                    if (trimmedToken.length() >= trimmedToken.indexOf("\\")) {
+                        final StringBuilder sb = new StringBuilder(trimmedToken);
+                        sb.deleteCharAt(trimmedToken.indexOf("\\"));
+                        trimmedToken = sb.toString();
+                        // when the trimmedToken has another \. meaning more to escape
+                        moreToEscape = trimmedToken.contains("\\") ? moreToEscape : false;
```
Comment (antares1470) [inline CoreUtilities/src/au/gov/asd/tac/constellation/utilities/query/QueryEvaluator.java:162]: change this to `moreToEscape = trimmedToken.contains("\\") && moreToEscape;`

File: CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/state/AnalyticViewState.java (hunk 1 above)
Comment (aldebaran30701) [reply CoreAnalyticView/src/au/gov/asd/tac/constellation/views/analyticview/state/AnalyticViewState.java]: Done

**Comment by aldebaran30701** [comment]
@antares1470 Much appreciated with the fast review - I've fixed up those points you outlined also.

File: CoreLayersView/nbproject/project.xml [hunk 7]
```diff
@@ -103,6 +111,15 @@
                     </run-dependency>
                 </dependency>
             </module-dependencies>
+            <test-dependencies>
+                <test-type>
+                    <name>unit</name>
+                    <test-dependency>
+                        <code-name-base>org.netbeans.libs.testng</code-name-base>
```
Comment (arcturus2) [inline CoreLayersView/nbproject/project.xml:118]: 😄

Comment (aldebaran30701) [reply CoreLayersView/nbproject/project.xml:118]: This is ok yes ? Just making sure it's just a laugh on behalf of tests 👍

File: CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java [hunk 8]
```diff
@@ -84,15 +75,108 @@ public void submit() {
         if (pane == null) {
             return;
         }
-
         final List<String> layerQueries = new ArrayList<>();
-        for (Node node : pane.getLayers().getChildren()) {
-            if (GridPane.getRowIndex(node) > 0 && GridPane.getColumnIndex(node) == 2) {
-                final TextArea queryTextArea = (TextArea) node;
-                layerQueries.add(queryTextArea.getText().isBlank() ? null : queryTextArea.getText());
+        for (LayerDescription layer : pane.getlayers()) { // TODO: This was the last change of tuesday
```
Comment (arcturus2) [inline CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:79]: I think only you will understand that TODO @aldebaran30701. Can you either elaborate or we get rid of it?

Comment (aldebaran30701) [reply CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewController.java:79]: Good catch - Should start searching the project for TODO: before committing

File: CoreLayersView/test/unit/src/au/gov/asd/tac/constellation/views/layers/utilities/UpdateElementBitmaskPluginNGTest.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        // Check Transaction set correctly
+        assertTrue(1 == graph.getIntValue(layerMaskT, txId1));
+        assertTrue(1.0f == graph.getFloatValue(layerVisibilityT, txId1));
+        assertTrue(false == graph.getBooleanValue(selectedT, txId1));
+                
+        assertTrue(1 == graph.getIntValue(layerMaskT, txId2));
+        assertTrue(1.0f == graph.getFloatValue(layerVisibilityT, txId2));
+        assertTrue(false == graph.getBooleanValue(selectedT, txId2));
+        
+        PluginExecution.withPlugin(new UpdateElementBitmaskPlugin(2,LayerAction.ADD)).executeNow(graph);
```
Comment (arcturus2) [inline CoreLayersView/test/unit/src/au/gov/asd/tac/constellation/views/layers/utilities/UpdateElementBitmaskPluginNGTest.java:116]: ❤️

**Comment by arcturus2** [comment]
@aldebaran30701  does this super seed #454 or does this get merged after it?

**Comment by aldebaran30701** [comment]
@arcturus2 
Merge in:
1. Layers View State #454 
2. Layers Nested Query rework #491 
3. Layers Context Menu (PR to come pretty soon)

I'm not 100% sure if all of the State fixes are live in #491 - a merge in the above order should be fine.

**Comment by arcturus2** [comment]
>Layers Context Menu (PR to come pretty soon)

Wow thanks !!

Cool @aldebaran30701 and I think I'd like to vote to promote this from Experimental to the Views menu very soon! Keep it up.

File: CoreLayersView/nbproject/project.xml (hunk 7 above)
Comment (arcturus2) [reply CoreLayersView/nbproject/project.xml:118]: Yeah its a happy laugh to see tests - keep it up!

**Comment by arcturus2** [comment]
@aldebaran30701 please advice when I can merge this into `master`

**Comment by aldebaran30701** [comment]
#524 now supersedes this

**Pull request closed without merging**